server/src/main/java/org/opensearch/tasks/TaskManager.java

```diff
@@ -116,10 +130,23 @@ public class TaskManager implements ClusterStateApplier {
     private final Map<TcpChannel, ChannelPendingTaskTracker> channelPendingTaskTrackers = ConcurrentCollections.newConcurrentMap();
     private final SetOnce<TaskCancellationService> cancellationService = new SetOnce<>();
 
-    public TaskManager(Settings settings, ThreadPool threadPool, Set<String> taskHeaders) {
+    private volatile boolean taskResourceConsumersEnabled;
+    private final Set<Consumer<Task>> taskResourceConsumer;
+
+    public TaskManager(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool, Set<String> taskHeaders) {
         this.threadPool = threadPool;
         this.taskHeaders = new ArrayList<>(taskHeaders);
         this.maxHeaderSize = SETTING_HTTP_MAX_HEADER_SIZE.get(settings);
+        this.taskResourceConsumersEnabled = TASK_RESOURCE_CONSUMERS_ENABLED.get(settings);
+        this.taskResourceConsumer = new HashSet<Consumer<Task>>() {
+            {
+                add(new TopNSearchTasksLogger(settings));
+            }
+        };
+
+        if (clusterSettings != null) {
+            clusterSettings.addSettingsUpdateConsumer(TASK_RESOURCE_CONSUMERS_ENABLED, this::setTaskResourceConsumersEnabled);
```

Comment:
You're leaking a reference to `this` in a constructor here. What this means is that clusterSettings gets a handle to a partially-constructed TaskManager instance. This is generally unsafe as it can violate thread safety guarantees (mostly involving `final` fields I think, because the compilers are free to reorder reads of final fields across synchronization barriers). Now, in this particular case, it's not likely to cause a problem because clusterSettings only gets a reference to a method that sets a volatile field, but it is still a pattern that I think should be avoided.

I would encourage you to change this, and it should be reasonably easy since you're already changing the signature of the only constructor of TaskManager. The safe way to do this is create a static factory method like the following:

```java
public static TaskManager createAndAddToSettings(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool, Set<String> taskHeaders) {
    final TaskManager taskManager = new TaskManager(settings, threadPool, taskHeaders);
    if (clusterSettings != null) {
        clusterSettings.addSettingsUpdateConsumer(TASK_RESOURCE_CONSUMERS_ENABLED, taskManager::setTaskResourceConsumersEnabled);
    }
}

private TaskManager(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool, Set<String> taskHeaders) {
    this.threadPool = threadPool;
    this.taskHeaders = new ArrayList<>(taskHeaders);
    this.maxHeaderSize = SETTING_HTTP_MAX_HEADER_SIZE.get(settings);
    this.taskResourceConsumersEnabled = TASK_RESOURCE_CONSUMERS_ENABLED.get(settings);
    this.taskResourceConsumer = new HashSet<Consumer<Task>>() {
        {
            add(new TopNSearchTasksLogger(settings));
        }
    };
}
```

Given the above code, I'd probably remove the null check on clusterSettings and make the plain constructor public for cases where you don't have a clusterSettings or add another plain `create()` static factory method. Also I'm not totally sold on the `createAndAddToSettings()` name so feel free to come up with something better there.

Reply:
Got it, I have added a static method ```createTaskManagerWithClusterSettings``` ) that TransportService. Retaining the plain constructor public for cases where we don't have a clusterSettings